Improve rustdoc layout

GuillaumeGomez opened the pull request:
This is an overtake of https://github.com/rust-lang/rust/pull/89385 originally written by @cynecx.

I kept the original commit and simply added the missing fixes into a new one. You can test it online [here](https://rustdoc.crud.net/imperio/improve-rustdoc-layout/std/index.html).

r? @jsha

rust-highfive commented:
Some changes occurred in HTML/CSS/JS.

cc @GuillaumeGomez

GuillaumeGomez commented:
I also renamed the `main` HTML ID into `main-content` to avoid confusion with the new `main` element. However it's a big change in term of line, so if preferred, I'll move it into another PR.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4356159011?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/633af0a3c7f82a6c18244096ae95f4e7a40de31e/checks/4356159011/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Finished release [optimized] target(s) in 27.90s
     Running unittests (build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/rustdoc-169b4e3f884f31b4)

running 76 tests
.............................................F.....F........................

---- html::markdown::tests::test_header_ids_multiple_blocks stdout ----
thread 'html::markdown::tests::test_header_ids_multiple_blocks' panicked at 'assertion failed: `(left == right)`
thread 'html::markdown::tests::test_header_ids_multiple_blocks' panicked at 'assertion failed: `(left == right)`
  left: `"<h2 id=\"main\" class=\"section-header\"><a href=\"#main\">Main</a></h2>"`,
 right: `"<h2 id=\"main-1\" class=\"section-header\"><a href=\"#main-1\">Main</a></h2>"`: original: # Main', src/librustdoc/html/markdown/tests.rs:202:9

---- html::markdown::tests::test_unique_id stdout ----
---- html::markdown::tests::test_unique_id stdout ----
thread 'html::markdown::tests::test_unique_id' panicked at 'assertion failed: `(left == right)`
  left: `["foo", "examples", "examples-1", "method.into_iter", "examples-2", "method.into_iter-1", "foo-1", "main", "search-1", "methods", "examples-3", "method.into_iter-2", "assoc_type.Item", "assoc_type.Item-1"]`,
 right: `["foo", "examples", "examples-1", "method.into_iter", "examples-2", "method.into_iter-1", "foo-1", "main-1", "search-1", "methods", "examples-3", "method.into_iter-2", "assoc_type.Item", "assoc_type.Item-1"]`', src/librustdoc/html/markdown/tests.rs:42:5

failures:
    html::markdown::tests::test_header_ids_multiple_blocks
    html::markdown::tests::test_unique_id
    html::markdown::tests::test_unique_id

test result: FAILED. 74 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s

error: test failed, to rerun pass '-p rustdoc --lib'


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--manifest-path" "/checkout/src/tools/rustdoc/Cargo.toml" "-p" "rustdoc:0.0.0" "--" "--quiet"


Build completed unsuccessfully in 0:28:55
```

</details>

cynecx commented on src/librustdoc/html/static/css/rustdoc.css:
```diff
@@ -285,28 +310,76 @@ li {
+	position: sticky;
+	min-width: 200px;
+	height: 100vh;
 }
```
Weird, something is missing here (Did something go wrong with your rebase?):

```css
top: 0;
left: 0;
```

Without it the sidebar is not sticky.

GuillaumeGomez replied:
Indeed, it has been moved in the next rule, which is invalid. Good catch!

jsha commented on src/librustdoc/html/static/css/rustdoc.css:
```diff
@@ -1646,21 +1695,62 @@ details.rustdoc-toggle[open] > summary.hideme::after {
+		padding-top: 0px;
+	}
+
+	.rustdoc {
```
As above this rule should go on `body`.

GuillaumeGomez replied:
Same answer as above. :-/

jsha commented on src/librustdoc/html/static/css/rustdoc.css:
```diff
@@ -1678,7 +1768,7 @@ details.rustdoc-toggle[open] > summary.hideme::after {
 	}
 
-	.sidebar .logo-container {
+	.rustdoc:not(.source) .sidebar .logo-container {
```
Seeing a bunch of `not()` selectors usually hints that we're missing some positive class that would be useful. For instance it looks like it would be useful to add a `doc` class to the body tag whenever we're not in source view. But I don't feel that's required as part of this PR.

jsha reviewed (changes requested):
Looking great! A few small nits and some naming discussion:

I find the hierarchy `main > .main-inner > #main-content` confusing; I would have expected `main-inner` to be the innermost. The primary function of `.main-inner` seems to be a place to hang `max-width: 960px` (for doc pages) or `max-width: unset` (for source pages). Maybe we can call it more precisely `.width-limiter` or something?

GuillaumeGomez commented on src/librustdoc/html/static/css/rustdoc.css:
```diff
@@ -248,6 +247,32 @@ textarea {
 
 /* end tweaks for normalize.css 8 */
 
+.rustdoc {
```
Actually we can't because of <docs.rs>. :-/

GuillaumeGomez commented on src/librustdoc/html/static/css/rustdoc.css:
```diff
@@ -285,28 +310,75 @@ li {
+
+.source .sidebar.expanded {
 	overflow-y: auto;
+	width: 300px !important;
```
Just checked and you're right, so removed.

GuillaumeGomez commented:
> Looking great! A few small nits and some naming discussion:
>
> I find the hierarchy `main > .main-inner > #main-content` confusing; I would have expected `main-inner` to be the innermost. The primary function of `.main-inner` seems to be a place to hang `max-width: 960px` (for doc pages) or `max-width: unset` (for source pages). Maybe we can call it more precisely `.width-limiter` or something?

Agreed, I added a commit for the renaming.

I also updated the files on the server so you can check too. :)

jsha commented:
@bors r+ rollup

bors commented:
:pushpin: Commit d7528e2157762fadb9665518fd1e4dee6d6a2809 has been approved by `jsha`

<!-- @bors r=jsha d7528e2157762fadb9665518fd1e4dee6d6a2809 -->
<!-- homu: {"type":"Approved","sha":"d7528e2157762fadb9665518fd1e4dee6d6a2809","approver":"jsha"} -->

bors commented:
:hourglass: Testing commit d7528e2157762fadb9665518fd1e4dee6d6a2809 with merge e2116acae59654bfab2a9729a024f3e2fd6d4b02...
<!-- homu: {"type":"BuildStarted","head_sha":"d7528e2157762fadb9665518fd1e4dee6d6a2809","merge_sha":"e2116acae59654bfab2a9729a024f3e2fd6d4b02"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4424214626?check_suite_focus=true)
Approved by: jsha
Pushing e2116acae59654bfab2a9729a024f3e2fd6d4b02 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jsha","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4424214626?check_suite_focus=true"},"merge_sha":"e2116acae59654bfab2a9729a024f3e2fd6d4b02"} -->

rust-timer commented:
Finished benchmarking commit (e2116acae59654bfab2a9729a024f3e2fd6d4b02): [comparison url](https://perf.rust-lang.org/compare.html?start=772d51f887fa407216860bf8ecf3f1a32fb795b4&end=e2116acae59654bfab2a9729a024f3e2fd6d4b02).

**Summary**: This change led to large relevant regressions 😿 in compiler performance.
- Large regression in instruction counts (up to 1.7% on `full` builds of `externs`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression

GuillaumeGomez commented:
There is basically no change in the rust source code so this is spurious.

@rustbot label: -perf-regression
